subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java

```diff
@@ -32,13 +32,6 @@ public RegistryAwareClassLoaderHierarchyHasher(ClassLoaderRegistry registry, Has
     private static Map<ClassLoader, String> collectKnownClassLoaders(ClassLoaderRegistry registry) {
         Map<ClassLoader, String> knownClassLoaders = Maps.newHashMap();
 
-        String javaVmVersion = String.format("%s|%s", System.getProperty("java.vm.name"), System.getProperty("java.vm.vendor"));
-        ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
-        if (systemClassLoader != null) {
-            addClassLoader(knownClassLoaders, systemClassLoader, "system-app" + javaVmVersion);
```

Comment:
That is strange. I think we should still add those two classloaders to the known classloaders, don't you think?

Reply:
What is a use case where we'd need them?

Reply:
I don't have a use-case. Though removing the classloaders here could be a breaking change, right? Do you know why they have been here in the first place?

Reply:
I don't have a clear memory, but here's my theory. This was added very early in the _task output caching_ project. We were very anxious about false positive cache hits, and it was unclear if there were significant differences in how JVMs worked wrt things like e.g. `HashMap` ordering or `hashCode()` implementations for standard library types etc. We wanted to err on the side of caution, but we also made a mistake, and never captured the Java version as originally intended. We also never checked (there is no test) if this piece of code actually had any effect.

As to whether it's okay to remove this... In short: I'm not 100% sure, though no test has failed, which is a good sign. For this piece of code to have any effect, you'd need to an input that is loaded via the system classloader, i.e. it has to have a type that is defined in the Java runtime library. The ones I can think of (`Integer`, `String`, `File`, `Map`) have special treatment as primitive-like types, and we never snapshot their `Class`. `Callable` might be an issue, but you never actually have a `Callable`, instead you have an implementation of `Callable` that _is_ loaded via the Gradle classloaders, so no problem there either. So all-in-all I expect that this code is never triggered and can be safely removed.

Reply:
I don't think we have such a type, because if we did, that test should have broke. What type should we add do you think? My main problem is that I don't really have a good idea for one.

Reply:
Types from the `java.time` package would be good candidates I think.

Reply:
Ah, like `java.time.Duration`. That makes sense. I think with the changes here we'd make a task having a property of type `Duration` be out-of-date all the time, since the classloader for that type is unknown after the change.

So I think a better change here would be to keep adding the classloaders, but set their hash to a constant, say, `Hashing.signature("SYSTEM")` or something.

In any case, let's add the test first, and then check if it indeed fails, and then fix it by re-adding the system classloaders with the fixed hash.

Reply:
@lptr I added a test just to check that a task having a property of type `Duration` is out-of-date all the time but it turns out it isn't

https://github.com/gradle/gradle/pull/20300/files#diff-97bc9f3d972bb9ec46c88ab98e683e8638a60555c4fec30295509a94bfd6e554R487

Reply:
We actually capture the classloader further down the line, but `Duration` has `null` (because Java thinks it's a primitive), and we later on ignore missing classpath hashes here:

https://github.com/gradle/gradle/blob/c7f67221c5cf025430542a2bf8e3ca6dae3a61e0/subprojects/model-core/src/main/java/org/gradle/internal/snapshot/impl/JavaSerializedValueSnapshot.java#L90-L98

Reply:
Update: it's not just for primitives. Here's `Class.getClassLoader()`'s Javadocs:

```java
    /**
     * Returns the class loader for the class.  Some implementations may use
     * null to represent the bootstrap class loader. This method will return
     * null in such implementations if this class was loaded by the bootstrap
     * class loader.
```

So the bootstrap classloader is always represented as `null`.

Reply:
Ok then, we have the test now so I suppose that is good enough.